crates/storage/provider/src/traits/stage_checkpoint.rs

```diff
@@ -6,4 +6,17 @@ use reth_primitives::stage::{StageCheckpoint, StageId};
 pub trait StageCheckpointProvider: Send + Sync {
     /// Fetch the checkpoint for the given stage.
     fn get_stage_checkpoint(&self, id: StageId) -> Result<Option<StageCheckpoint>>;
+
+    /// Get stage checkpoint progress.
+    fn get_stage_checkpoint_progress(&self, id: StageId) -> Result<Option<Vec<u8>>>;
+}
+
+/// The trait for updating stage checkpoint related data.
+#[auto_impl::auto_impl(&, Arc)]
+pub trait StageCheckpointWriter: Send + Sync {
+    /// Save stage checkpoint.
+    fn save_stage_checkpoint(&self, id: StageId, checkpoint: StageCheckpoint) -> Result<()>;
+
+    /// Save stage checkpoint progress.
+    fn save_stage_checkpoint_progress(&self, id: StageId, checkpoint: Vec<u8>) -> Result<()>;
```

Comment:
I like this

Reply:
unrelated to this PR changes: I think I'd be confused what does it mean if I don't look in which table it eventually gets written. We currently have these weird leftover of Merkle stage checkpoint logic which still uses the `SyncStageProgress` table instead of saving into the `SyncStage` as e.g. hashing stage do. That's because Merkle checkpoint is actually pretty heavy compared to hashing stages', so it can't be `Copy`, hence we needed to make `(Exec|Unwind)(Input|Output)` non-`Copy` too.

Will create an issue to address it in the future, as it's purely a thing of convenience which doesn't affect performance in any way.